src/main/java/az/ingress/mscategory/service/concrete/CategoryServiceHandler.java

```diff
@@ -0,0 +1,177 @@
+import static az.ingress.mscategory.model.enums.CategoryStatus.ACTIVE;
+import static az.ingress.mscategory.model.enums.CategoryStatus.DELETED;
+
+@Log
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceHandler implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+    private final CacheUtil cacheUtil;
+    private static final String CACHE_KEY = "categories";
+
+    @Override
+    public List<CategoryResponse> getCategories() {
+        var categoryEntities = fetchAllCategories();
+        var categoryMap = createCategoryMap(categoryEntities);
+        return buildCategoryHierarchy(categoryEntities, categoryMap);
+    }
+
+    private Iterable<CategoryEntity> fetchAllCategories() {
+        Iterable<CategoryEntity> cachedCategories = cacheUtil.getBucket(CACHE_KEY);
+        if (cachedCategories != null) {
+            return cachedCategories;
+        }
+        var categoryEntities = categoryRepository.findAll();
+        cacheUtil.saveToCache(CACHE_KEY, categoryEntities, 1L, ChronoUnit.DAYS);
+        return categoryEntities;
+    }
+
+    public Map<Long, CategoryResponse> createCategoryMap(Iterable<CategoryEntity> categoryEntities) {
+        Map<Long, CategoryResponse> categoryMap = new HashMap<>();
+        for (var categoryEntity : categoryEntities) {
+            var categoryResponse = new CategoryResponse();
+            categoryResponse.setId(categoryEntity.getId());
+            categoryResponse.setParentId(categoryEntity.getId());
+            categoryResponse.setName(categoryEntity.getName());
+            categoryResponse.setChildren(new ArrayList<>());
+            categoryMap.put(categoryEntity.getId(), categoryResponse);
+        }
+        return categoryMap;
+    }
+
+    private List<CategoryResponse> buildCategoryHierarchy(Iterable<CategoryEntity> categoryEntities, Map<Long, CategoryResponse> categoryMap) {
+        List<CategoryResponse> rootCategories = new ArrayList<>();
+        for (var categoryEntity : categoryEntities) {
+            if (categoryEntity.getParentId() == null) {
+                rootCategories.add(categoryMap.get(categoryEntity.getId()));
+            } else {
+                addSubCategoryToParent(categoryMap, categoryEntity);
+            }
+        }
+        return rootCategories;
+    }
+
+    private void addSubCategoryToParent(Map<Long, CategoryResponse> categoryMap, CategoryEntity categoryEntity) {
+        var parentCategory = categoryMap.get(categoryEntity.getParentId());
+        if (parentCategory != null) {
+            parentCategory.getChildren().add(categoryMap.get(categoryEntity.getId()));
+        }
+    }
+
+    @Transactional
+    @Override
+    public List<CategoryResponse> createCategory(CategoryRequest categoryRequest) {
+        List<CategoryResponse> categoryResponses = new ArrayList<>();
+        for (var categoryDetail : categoryRequest.getCategories()) {
```

Comment:
change to stream api, move to saveAll, move to updateCategoriesInCache to async with retryable